### PR TITLE
Record max call elapsed times for single methods

### PR DESCRIPTION
**Pull request by josh-127**
# Description

This will record worst-case timings for each traced method call (also called "maximum call time"). This is useful for keeping track of whether occassional bottlenecks occur in a method. I decided to display maximum call times as a column, thus allowing the user to sort traces by this new field.

### CODE REVIEW

File: src/test/java/com/google/idea/perf/CallTreeTest.kt [hunk 1]
```diff
@@ -25,11 +25,17 @@ class CallTreeTest {
...
         override val wallTime: Long,
+        override val maxCallTime: Long,
         childrenList: List<Tree> = emptyList()
     ) : CallTree {
         override val children = childrenList.associateBy { it.tracepoint }
     }
 
+    init {
+        // Enforce LF line endings for developers on Windows.
+        System.setProperty("line.separator", "\n");
```
Comment (gharrma) [inline src/test/java/com/google/idea/perf/CallTreeTest.kt:36]: I guess this will affect other tests too, since the property persists for the life of the JVM.
Maybe it's harmless, although I also wonder if there's another way.
Not worth worrying about now though!

File: src/main/java/com/google/idea/perf/CallTree.kt [hunk 2]
```diff
@@ -21,6 +21,7 @@ interface CallTree {
     val tracepoint: Tracepoint
     val callCount: Long
     val wallTime: Long
+    val maxCallTime: Long
```
Comment (gharrma) [inline src/main/java/com/google/idea/perf/CallTree.kt:24]: Note for the future: at some point we might want to measure 'CPU time' in addition to wall time, at which point we might want to rename this to 'maxWallTime'.

File: src/main/java/com/google/idea/perf/CallTreeBuilder.kt [hunk 3]
```diff
@@ -97,7 +101,9 @@ class CallTreeBuilder(
         val now = clock.sample()
         val stack = generateSequence(currentNode, Tree::parent).toList().asReversed()
         for (node in stack) {
-            node.wallTime += now - node.startWallTime
+            val elapsedTime = now - node.startWallTime
+            node.wallTime += elapsedTime
+            node.maxCallTime = node.maxCallTime.coerceAtLeast(elapsedTime)
             node.startWallTime = now
```
Comment (gharrma) [inline src/main/java/com/google/idea/perf/CallTreeBuilder.kt:107]: Currently we set `node.startWallTime = now` when `buildAndReset()` is called.

This is fine for `wallTime` because it is accumulated along the way.  
However, it poses an issue for the `maxCallTime` calculation.

Comment (josh-127) [reply src/main/java/com/google/idea/perf/CallTreeBuilder.kt:107]: Good catch! I fixed this issue in an upcoming commit, but I have one question. What would the root node's max wall time be while the call stack is not empty?

Comment (gharrma) [reply src/main/java/com/google/idea/perf/CallTreeBuilder.kt:107]: I think there are a couple reasonable options

1. "max time" could consider calls still on the stack. So if `foo()` is still on the stack, and it has been on the stack for `400 ms`, then the max time for `foo()` should be at least `400 ms`.

2. Or, "max time" could be computed only from completed calls. In that case the "max time" for the root node would always be 0 (since it's always on the stack). That's OK because the root node is not a user-visible node anyway.

Either option is fine with me!

Comment (josh-127) [reply src/main/java/com/google/idea/perf/CallTreeBuilder.kt:107]: Alright thanks! To stay consistent with the wall time's update rate, I decided to pick option 1. In this case, the root node will always increase to satisfy an invariant (it never pops), but it's more of an emergent property rather than a corner case.

File: src/test/java/com/google/idea/perf/CallTreeTest.kt (hunk 1 above)
Comment (josh-127) [reply src/test/java/com/google/idea/perf/CallTreeTest.kt:36]: Sure! I'll keep an eye on it whenever I start making new unit tests.

File: src/main/java/com/google/idea/perf/CallTree.kt (hunk 2 above)
Comment (josh-127) [reply src/main/java/com/google/idea/perf/CallTree.kt]: Yeah, good idea. I decided to rename it as soon as possible. Maybe we should also rename `wallTime` to `totalWallTime` later on. What do you think?

Comment (gharrma) [reply src/main/java/com/google/idea/perf/CallTree.kt]: Yeah, `wallTime` or `totalWallTime` both sound good to me. Let's keep `wallTime` for now.
